File: 2020/day5/day5.js
const fs = require('fs');
const path = require('path');

function parseTickets(tickets) {
  const ROWS = 128;
  const COLS = 8;
  const seatIds = [];

  for (const ticket of tickets) {
    const chars = ticket.split('');
    let front = 0;
    let back = ROWS - 1;
    let left = 0;
    let right = COLS - 1;
    for (const char of chars) {
      let row = back - front;
      let col = right - left;
      if (char === 'F') {
        back -= Math.ceil(row / 2);
      } else if (char === 'B') {
        front += Math.ceil(row / 2);
      } else if (char === 'L') {
        right -= Math.ceil(col / 2);
      } else if (char === 'R') {
        left += Math.ceil(col / 2);
      }
    }
    const seatId = (front * COLS) + right;
    seatIds.push(seatId);
  }
  return seatIds;
}

function part1(input) {
  const tickets = input
    .trim()
    .split('\n');

  const seatIds = parseTickets(tickets);

  return seatIds.reduce((cur, acc) => {
    if (cur > acc) acc = cur;
    return acc;
  });
}

function part2(input) {
  const tickets = input
    .trim()
    .split('\n');

  const seatIds = parseTickets(tickets);

  const maxId = seatIds.reduce((cur, acc) => {
    if (cur > acc) acc = cur;
    return acc;
  });
  for (let i = 1; i < maxId - 1; i++) {
    if (!seatIds.includes(i) &&
      seatIds.includes(i + 1) &&
      seatIds.includes(i - 1)) {
      return i;
    }
  }
  return -1;
}

const input = fs.readFileSync(path.resolve(__dirname, './input.txt'), 'utf8');
console.log('day5 part1:', part1(input));
console.log('day5 part2:', part2(input));

/*
const input = require('../input');
const log = console.log;
const compose = (...fns) => (args) => fns.reduceRight((arg, fn) => fn(arg), args);
const replace = (...args) => (string) => string.replace(...args);

const replaceUpperHalfsWithOne = replace(/[BR]/g, 1);
const replaceLowerHalfsWithZero = replace(/[FL]/g, 0);
const toIntFromBinaryString = (string) => parseInt(string, 2);
const toSeatID = compose(toIntFromBinaryString, replaceLowerHalfsWithZero, replaceUpperHalfsWithOne);

const ascending = (a, b) => a - b;
const isMissingSeatBehind = (id, _, ids) => !ids.includes(id - 1) && ids.includes(id - 2);

const seatIDs = input(__dirname, '/input.txt').split('\n').map(toSeatID);

const solutionOne = Math.max(...seatIDs);
log(`Solution p1. ${solutionOne}`);

const seatInFrontOfOwn = seatIDs.sort(ascending).find(isMissingSeatBehind);
const solutionTwo = seatInFrontOfOwn - 1;
log(`Solution p2. ${solutionTwo}`);
*/

File: 2020/day5/day5.clj
'{:dependencies [[org.clojure/clojure "1.10.1"]]}
(require '[clojure.string :as str])

(def input
  (->>
   (slurp "2020/day5/input.txt")
   (str/split-lines)))

(defn includes? [list val] (contains? (set list) val))

(defn seat-id [string]
  (->>
   (seq string)
   (reduce (fn [[front back left right] c]
             (case c
               \B [(+ front (Math/ceil (/ (- back front) 2))) back left right]
               \F [front (- back (Math/ceil (/ (- back front) 2))) left right]
               \R [front back (+ left (Math/ceil (/ (- right left) 2))) right]
               \L [front back left (- right (Math/ceil (/ (- right left) 2)))])) [0 127 0 7])
   ((fn [[front _ left _]] (+ (* front 8) left)))
   (int)))

(defn part1 []
  (prn "day5 part1:"
       (->>
        input
        (map seat-id)
        (apply max))))

(defn part2 []
  (prn "day5 part2:"
       (->>
        input
        (#(let [seat-ids (map seat-id %)
                max-id (apply max seat-ids)]
            (for [i (range 1 max-id)
                  :when (and  (not (includes? seat-ids i))
                              (includes? seat-ids (inc i))
                              (includes? seat-ids (dec i)))] i)))
        (first))))

(part1)
(part2)

;; (ns aoc.2020.d05
;;   (:require [aoc.file-util :as file-util]
;;             [clojure.string :as str]))

;; (def input (file-util/read-lines "2020/d05.txt"))

;; (defn seat-id [seat-code]
;;   (-> seat-code
;;       (str/escape {\B 1, \F 0, \R 1, \L 0})
;;       (Integer/parseInt 2)))

;; (defn part-1 [input] (apply max (map seat-id input)))

;; (defn part-2 [input]
;;   (let [taken-seats (sort (map seat-id input))
;;         all-seats (range (first taken-seats) (inc (last taken-seats)))]
;;     (first (clojure.set/difference (set all-seats) (set taken-seats)))))
